src/esm.ts

```diff
@@ -48,21 +49,35 @@ function _genStatement (type: 'import' | 'export' | 'import type' | 'export type
 
   const namesStr = _names.map(i => i.as ? `${i.name} as ${i.as}` : i.name).join(', ')
   if (nameArray) {
-    return `${type} { ${namesStr} } from ${genString(specifier, opts)};`
+    return `${type} { ${namesStr} } from ${genString(specifier, opts)}${_getAssertClause(type, opts.assert)};`
   }
-  return `${type} ${namesStr} from ${genString(specifier, opts)};`
+  return `${type} ${namesStr} from ${genString(specifier, opts)}${_getAssertClause(type, opts.assert)};`
+}
+
+function _getAssertClause (type: ImportExportType, assert?: { type: string }) {
+  // import assertions isn't specified type-only import or export on Typescript
+  if (type === 'import type' || type === 'export type') {
+    return ''
+  }
+  return assert ? ` assert { type: ${genString(assert.type)} }` : ''
 }
 
 export interface DynamicImportOptions extends CodegenOptions {
   comment?: string
   wrapper?: boolean
-  interopDefault?: boolean
+  interopDefault?: boolean,
+  // https://github.com/tc39/proposal-import-assertions
+  // https://tc39.es/proposal-import-assertions/
+  assert?: {
+    type: string
+  }
 }
 export function genDynamicImport (specifier: string, opts: DynamicImportOptions = {}) {
   const commentStr = opts.comment ? ` /* ${opts.comment} */` : ''
   const wrapperStr = (opts.wrapper === false) ? '' : '() => '
   const ineropStr = opts.interopDefault ? '.then(m => m.default || m)' : ''
-  return `${wrapperStr}import(${genString(specifier, opts)}${commentStr})${ineropStr}`
+  const assertStr = opts.assert ? `, { assert: { type: ${genString(opts.assert.type)} } }` : ''
```

Comment:
We need to check for `opts.asserts?.type`

Reply:
In tc39 proposal and github discussion, I seem that import assertions `type` specifications still don't decide. 
https://github.com/tc39/proposal-import-assertions#proposed-semantics-and-interoperability

so I don't think we need to check it.

The generated code is processed by bundler loader / plugin. I believe that the `type` is the responsibility of the user.

Reply:
Indeed. But users might (mistakely provide something like `assert: {}` or `assert { anotherKey: true}`. In this case, we will generate invalid code if `assert.type` is not specified.

Reply:
I've fixed it!
https://github.com/unjs/knitwork/pull/34/commits/a8307b45127145ab527079d15755ad0af093eb72